### PR TITLE
Add BFF proxy sample

### PR DESCRIPTION
**Pull request by derisen**
## Purpose

* Adds a new sample implementing the [BFF Proxy](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-browser-based-apps#name-backend-for-frontend-bff-pr) pattern using the Visual Studio ASP.NET Core with React.js template

## Does this introduce a breaking change?

```
[ ] Yes
[ x ] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ x ] Other... Please describe: new sample
```

## How to Test

*  Get the code

```console
    git clone https://github.com/Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2.git
    git checkout bff-proxy-sample
    cd  ./2-WebApp-graph-user/2-6-BFF-Proxy
    dotnet restore
```

* Test the code

```
dotnet run
```

## What to Check

| Do | Verify |
|-----|-----|
| Click sign-in | Confirm that the user navigates to AAD login and ASP.NET auth cookie is placed in cookie storage afterwards |
| Click sign-out |  Confirm that the user navigates to AAD logout and the ASP.NET auth cookie is deleted afterwards |
| Restart the app, sign-in, shut down the app, then launch the app again | Confirm that the cookie in storage is ignored and the app attempts to login again |
| Launch the app, navigate to `/fetch-graph` without signing in first |  Confirm that the app redirects to AAD login |
| Launch the app, sign-in, click call Graph | Confirm that the app fetches users Graph profile |
| Trigger CAE (e.g. end user session on AAD portal), navigate to `/fetch-graph` | Confirm that the app prompts user for challenge |

### CODE REVIEW

File: 2-WebApp-graph-user/2-6-BFF-Proxy/AppCreationScripts/sample.json [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    "AADApps": [
+        {
+            "Id": "client",
+            "Name": "CallGraphBFF",
+            "Kind": "WebApp",
+            "Audience": "AzureADMyOrg",
+            "HomePage": "https://localhost:7000",
+            "ReplyUrls": "https://localhost:7000/api/auth/signin-oidc, https://localhost:7000/api/auth/signout-oidc",
+            "SDK": "MicrosoftIdentityWeb",
+            "SampleSubPath": "2-WebApp-graph-user\\2-6-BFF-Proxy",
```
Comment (salman90) [inline 2-WebApp-graph-user/2-6-BFF-Proxy/AppCreationScripts/sample.json:30]:
```suggestion
            "SampleSubPath": "2-WebApp-graph-user\\2-6-BFF-Proxy\\CallGraphBFF",
```

File: 2-WebApp-graph-user/2-6-BFF-Proxy/CallGraphBFF/Controllers/AuthController.cs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        return new SignOutResult(optionList, new AuthenticationProperties
+        {
+            RedirectUri = "/"
+        });
+    }
+
+    [HttpGet("account")]
+    public ActionResult GetAccount()
+    {
+        if (User.Identity.IsAuthenticated)
```
Comment (v-michaelmi) [inline 2-WebApp-graph-user/2-6-BFF-Proxy/CallGraphBFF/Controllers/AuthController.cs:41]: You can resolve this issue with:

```
if (User.Identity is not null && User.Identity.IsAuthenticated)
```

**Comment by v-michaelmi** [comment]
Couple of nits but otherwise LGTM

File: 2-WebApp-graph-user/2-1-Call-MSGraph/AppCreationScripts/Cleanup-WithCertificates.ps1 [hunk 3]
```diff
@@ -0,0 +1,154 @@
+﻿
```
Comment (salman90) [inline 2-WebApp-graph-user/2-1-Call-MSGraph/AppCreationScripts/Cleanup-WithCertificates.ps1:1]: Did you want this PR to update 2-1 sample?

Comment (derisen) [reply 2-WebApp-graph-user/2-1-Call-MSGraph/AppCreationScripts/Cleanup-WithCertificates.ps1:1]: Ah, this is accidental -will revert back

File: 2-WebApp-graph-user/2-6-BFF-Proxy/ReadmeFiles/sequence.png
Comment (derisen) [inline 2-WebApp-graph-user/2-6-BFF-Proxy/ReadmeFiles/sequence.png:1]: - typo: login.microsoftonline.com
- Also, GitHub can render mermaid diagrams -replace the pic

File: 2-WebApp-graph-user/2-6-BFF-Proxy/README.md [hunk 4]
```diff
@@ -0,0 +1,417 @@
...
+* [Explore the sample](#explore-the-sample)
+* [Troubleshooting](#troubleshooting)
+* [About the code](#about-the-code)
+* [Next Steps](#next-steps)
+* [Contributing](#contributing)
+* [Learn More](#learn-more)
+
+## Overview
+
+This sample demonstrates a React single-page application (SPA) with an ASP.NET Core backend that authenticates users and calls the Microsoft Graph API using the backend for frontend (BFF) proxy architecture. In this architecture, access tokens are retrieved and stored within the secure backend context, and the client side JavaScript application, which is served by the ASP.NET web app, is only indirectly involved in the authN/authZ process by routing the token and API requests to the backend. The trust between the frontend and backend is established via a secure cookie upon successful sign-in.
```
Comment (derisen) [inline 2-WebApp-graph-user/2-6-BFF-Proxy/README.md:38]: reminder: link to ietf-draft for "BFF proxy architecture" https://datatracker.ietf.org/doc/html/draft-ietf-oauth-browser-based-apps#section-6.2

File: 2-WebApp-graph-user/2-6-BFF-Proxy/CallGraphBFF/Program.cs [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    options.Cookie.SameSite = SameSiteMode.Strict;
+    options.Cookie.SecurePolicy = CookieSecurePolicy.Always;
+    options.Cookie.HttpOnly = true;
+    options.Cookie.IsEssential = true;
+
+    options.Events = new CustomCookieAuthenticationEvents(); // modifies the behavior of certain cookie authentication events.
+});
+
+builder.Services.AddControllersWithViews()
+    .AddMicrosoftIdentityUI();
```
Comment (derisen) [inline 2-WebApp-graph-user/2-6-BFF-Proxy/CallGraphBFF/Program.cs:42]: reminder: AddMicrosoftIdentityUI here seems superflous

File: 2-WebApp-graph-user/2-6-BFF-Proxy/CallGraphBFF/Utils/CustomCookieAuthenticationEvents.cs [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    public override Task RedirectToAccessDenied(RedirectContext<CookieAuthenticationOptions> context)
+    {
+        context.Response.StatusCode = (int)HttpStatusCode.Unauthorized;
+        return Task.CompletedTask;
+    }
+
+    // Reject authentication cookie if no tokens are found in the cache
+    public async override Task ValidatePrincipal(CookieValidatePrincipalContext context)
+    {
+        try
```
Comment (jmprieur) [inline 2-WebApp-graph-user/2-6-BFF-Proxy/CallGraphBFF/Utils/CustomCookieAuthenticationEvents.cs:25]: that's a great experience. 
@jennyf19 : would solve https://github.com/AzureAD/microsoft-identity-web/issues/13  elegantly

File: 2-WebApp-graph-user/2-6-BFF-Proxy/CallGraphBFF/appsettings.json [hunk 7]
```diff
@@ -0,0 +1,30 @@
+{
+  "AzureAd": {
+    "Instance": "https://login.microsoftonline.com/",
+    "TenantId": "[Enter 'common', or 'organizations' or the Tenant Id (obtained from the Azure portal. Select 'Endpoints' from the 'App registrations' blade and use the GUID in any of the URLs), e.g. da41245a5-11b3-996c-00a8-4d99re19f292]",
+    "ClientId": "[Enter the Client Id (Application ID obtained from the Azure portal), e.g. ba74781c2-53c2-442a-97c2-3d60re42f403]",
+    "ClientSecret": "[Copy the client secret added to the app from the Azure portal]",
```
Comment (jmprieur) [inline 2-WebApp-graph-user/2-6-BFF-Proxy/CallGraphBFF/appsettings.json:6]: We could use the ClientCredentials collection insted of ClientSecret and ClientCertificates
(docs still being written though)

File: 2-WebApp-graph-user/2-6-BFF-Proxy/AppCreationScripts/AppCreationScripts.md [hunk 8]
```diff
@@ -0,0 +1,138 @@
+# Registering sample apps with the Microsoft identity platform and updating configuration files using PowerShell
+
+## Overview
+
+### Quick summary
+
+1. Run the script to create your Azure AD application and configure the code of the sample application accordingly.
+
```
[the remark this answers is not in the document]
Comment (jmprieur) [reply 2-WebApp-graph-user/2-6-BFF-Proxy/AppCreationScripts/AppCreationScripts.md:8]: it should also be at the top of the scripts themselves are comments

**Review by jmprieur — COMMENTED** [review]
LGTM
I've a few questions

File: 2-WebApp-graph-user/2-1-Call-MSGraph/AppCreationScripts/Cleanup-WithCertificates.ps1 (hunk 3 above)
Comment (derisen) [reply 2-WebApp-graph-user/2-1-Call-MSGraph/AppCreationScripts/Cleanup-WithCertificates.ps1]: done

File: 2-WebApp-graph-user/2-6-BFF-Proxy/CallGraphBFF/Program.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (derisen) [reply 2-WebApp-graph-user/2-6-BFF-Proxy/CallGraphBFF/Program.cs]: You're right, `AddMicrosoftIdentityUI` is not used here. I've removed it, but not sure how to go about combining the controllers..

File: 2-WebApp-graph-user/2-6-BFF-Proxy/CallGraphBFF/appsettings.json (hunk 7 above)
Comment (derisen) [reply 2-WebApp-graph-user/2-6-BFF-Proxy/CallGraphBFF/appsettings.json]: done! (sorry for taking so long to get back to this)

File: 2-WebApp-graph-user/2-6-BFF-Proxy/AppCreationScripts/AppCreationScripts.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (derisen) [reply 2-WebApp-graph-user/2-6-BFF-Proxy/AppCreationScripts/AppCreationScripts.md:8]: Aah, we were told this won't be needed anymore with PS v7. Do you think we should revert back (have you tried with v7?)

File: 2-WebApp-graph-user/2-6-BFF-Proxy/README.md (hunk 4 above)
Comment (derisen) [reply 2-WebApp-graph-user/2-6-BFF-Proxy/README.md]: done

File: 2-WebApp-graph-user/2-6-BFF-Proxy/AppCreationScripts/AppCreationScripts.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (jmprieur) [reply 2-WebApp-graph-user/2-6-BFF-Proxy/AppCreationScripts/AppCreationScripts.md:8]: yes. they are totally needed.

**Review by jmprieur — APPROVED** [review]
LGTM
Thanks @derisen

File: 2-WebApp-graph-user/2-6-BFF-Proxy/AppCreationScripts/AppCreationScripts.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (derisen) [reply 2-WebApp-graph-user/2-6-BFF-Proxy/AppCreationScripts/AppCreationScripts.md:14]: done! I've also created an issue for the codegen to add this instruction back https://github.com/Azure-Samples/GenerateScriptsForSample/issues/102

**Pull request merged**